java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java

```diff
@@ -0,0 +1,123 @@
+import org.sonar.java.resolve.Symbol;
+import org.sonar.java.resolve.Symbol.MethodSymbol;
+import org.sonar.java.resolve.Symbol.TypeSymbol;
+import org.sonar.java.resolve.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S2160",
+  name = "Subclasses that add fields should override \"equals\"",
+  tags = {"bug"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("30min")
+public class EqualsNotOverriddenInSubclassCheck extends SubscriptionBaseVisitor {
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+    if (hasSemantic()
+      && hasAtLeastOneField(classTree)
+      && !implementsEquals(classTree)
+      && parentClassImplementsEquals(classTree)) {
+      addIssue(classTree, "Override this superclass' \"equals\" method.");
+    }
+  }
+
+  private boolean hasAtLeastOneField(ClassTree classTree) {
+    for (Tree member : classTree.members()) {
+      if (isField(member)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isField(Tree tree) {
+    return tree.is(Kind.VARIABLE) && !((VariableTree) tree).modifiers().modifiers().containsAll(Lists.newArrayList(Modifier.FINAL, Modifier.STATIC));
+  }
+
+  private boolean implementsEquals(ClassTree classTree) {
+    return hasEqualsMethod(((ClassTreeImpl) classTree).getSymbol());
+  }
+
+  private boolean parentClassImplementsEquals(ClassTree tree) {
+    Tree superClass = tree.superClass();
+    if (superClass != null) {
+      TypeSymbol superClassSymbol = ((AbstractTypedTree) superClass).getSymbolType().getSymbol();
+      while (!superClassSymbol.getType().is("java.lang.Object")) {
+        if (hasEqualsMethod(superClassSymbol)) {
+          return true;
+        }
+        Type superClassType = superClassSymbol.getSuperclass();
+        if (superClassType == null) {
+          break;
+        }
+        superClassSymbol = superClassType.getSymbol();
+      }
+    }
+    return false;
+  }
+
+  private boolean hasEqualsMethod(TypeSymbol superClassSymbol) {
+    List<Symbol> equalsMembers = superClassSymbol.members().lookup("equals");
+    for (Symbol symbol : equalsMembers) {
+      if (isEqualsMethod(symbol)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isEqualsMethod(Symbol symbol) {
```

Comment:
I am pretty sure there is already an implementation somewhere of this method. This should probably be factorised somewhere (probably in Symbol class)